ebpf/disk-failure/injection.bpf.c

```diff
@@ -0,0 +1,100 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2023 Datadog, Inc.
+
+// +build ignore
+#include "injection.bpf.h"
+
+const volatile pid_t target_pid = 0;
+const volatile pid_t exclude_pid;
+const volatile char filter_path[61];
+
+struct data_t {
+    u32 ppid;
+    u32 pid;
+    u32 tid; 
+    u32 id;
+    char comm[100];
+};
+
+struct {
+    __uint(type, BPF_MAP_TYPE_PERF_EVENT_ARRAY);
+    __uint(max_entries, 1024);
+    __type(key, int);
+    __type(value, u32);
+} events SEC(".maps");
+
+SEC("kprobe/sys_openat")
+int injection_disk_failure(struct pt_regs *ctx)
+{
+    struct data_t data = {};
+
+    // Get data of the current process
+    u32 ppid = 0;
+    u32 pid = bpf_get_current_pid_tgid();
+    if (pid == exclude_pid) {
+        return 0;
+    }
+    u32 tid = bpf_get_current_pid_tgid() >> 32;
+    u32 gid = bpf_get_current_uid_gid();
+
+    if (pid != 1) {
+        // Get parent pid
+        struct task_struct *task;
+        struct task_struct *real_parent;
+        task = (struct task_struct *)bpf_get_current_task();
+        bpf_probe_read(&real_parent, sizeof(real_parent), &task->real_parent);
+        bpf_probe_read(&ppid, sizeof(ppid), &real_parent->tgid);
+
+        // Allow only children and parent process.
+        if (target_pid != 0 && ppid != target_pid && pid != target_pid) {
+          return 0;
+        }
+    }
+
+    if (ppid == exclude_pid || tid == exclude_pid) {
+        return 0;
+    }
+
+// Exclude this part of code if the following variables are not defined.
+// It allows the go program to compile without error.
+#if defined(__TARGET_ARCH_arm64) || defined(__TARGET_ARCH_x86)
+    // Allow only file with the desired prefix.
+    struct pt_regs *real_regs = (struct pt_regs *)PT_REGS_PARM1(ctx);
+    char *path = (char *)PT_REGS_PARM2_CORE(real_regs);
+    char cmp_path_name[62];
+    bpf_probe_read(&cmp_path_name, sizeof(cmp_path_name), path);
+    char cmp_expected_path[62];
+    bpf_probe_read(cmp_expected_path, sizeof(cmp_expected_path), filter_path);
+    int filter_len = (int) (sizeof(filter_path) / sizeof(filter_path[0])) - 1;
+   
+    if (filter_len > 62) {
```

Comment:
we should probably have the 62 as a const here

Reply:
The first time I have tried, but the Linux kernel returns an error. So, I prefer to remove the constant. I understand it is not clean, but I don't find a way to make it work with a constant. I will create a constant and run the program to send you the error.

Reply:
With the usage of the `path_max_length` constant:
``` go
const int path_max_length = 62;
...
    char cmp_path_name[path_max_length];
    bpf_probe_read(&cmp_path_name, sizeof(cmp_path_name), path);
    char cmp_expected_path[path_max_length];
    bpf_probe_read(cmp_expected_path, sizeof(cmp_expected_path), filter_path);
    int filter_len = (int) (sizeof(filter_path) / sizeof(filter_path[0])) - 1;
   
    if (filter_len > path_max_length) {
        return 0;
    }
...
```
I have the following error during the compilation:
<detail>
<p>

```
fatal error: error in backend: Unsupported dynamic stack allocation
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace, preprocessed source, and associated run script.
Stack dump:
0.      Program arguments: clang -D __TARGET_ARCH_arm64 -I /usr/include/aarch64-linux-gnu -O2 -c -target bpf -g -o ../bin/injector/ebpf/bpf-disk-failure.bpf.o disk-failure/injection.bpf.c
1.      <eof> parser at end of file
2.      Code generation
3.      Running pass 'Function Pass Manager' on module 'disk-failure/injection.bpf.c'.
4.      Running pass 'BPF DAG->DAG Pattern Instruction Selection' on function '@injection_disk_failure'
 #0 0x0000ffff7e8643c4 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) (/usr/lib/llvm-15/bin/../lib/libLLVM-15.so.1+0xef43c4)
 #1 0x0000ffff7e862418 llvm::sys::RunSignalHandlers() (/usr/lib/llvm-15/bin/../lib/libLLVM-15.so.1+0xef2418)
 #2 0x0000ffff7e78c814 (/usr/lib/llvm-15/bin/../lib/libLLVM-15.so.1+0xe1c814)
 #3 0x0000ffff7e78c79c (/usr/lib/llvm-15/bin/../lib/libLLVM-15.so.1+0xe1c79c)
 #4 0x0000ffff7e85eb64 llvm::sys::Process::Exit(int, bool) (/usr/lib/llvm-15/bin/../lib/libLLVM-15.so.1+0xeeeb64)
 #5 0x0000aaaaca9833a0 (/usr/lib/llvm-15/bin/clang+0x133a0)
 #6 0x0000ffff7e79dea0 llvm::report_fatal_error(llvm::Twine const&, bool) (/usr/lib/llvm-15/bin/../lib/libLLVM-15.so.1+0xe2dea0)
 #7 0x0000ffff7e79dd84 llvm::report_fatal_error(llvm::Twine const&, bool) (/usr/lib/llvm-15/bin/../lib/libLLVM-15.so.1+0xe2dd84)
 #8 0x0000ffff808bf708 (/usr/lib/llvm-15/bin/../lib/libLLVM-15.so.1+0x2f4f708)
 #9 0x0000ffff7eed695c (/usr/lib/llvm-15/bin/../lib/libLLVM-15.so.1+0x156695c)
#10 0x0000ffff7eed60f4 llvm::SelectionDAG::Legalize() (/usr/lib/llvm-15/bin/../lib/libLLVM-15.so.1+0x15660f4)
#11 0x0000ffff7efebc18 llvm::SelectionDAGISel::CodeGenAndEmitDAG() (/usr/lib/llvm-15/bin/../lib/libLLVM-15.so.1+0x167bc18)
#12 0x0000ffff7efeb2ec llvm::SelectionDAGISel::SelectAllBasicBlocks(llvm::Function const&) (/usr/lib/llvm-15/bin/../lib/libLLVM-15.so.1+0x167b2ec)
#13 0x0000ffff7efe943c llvm::SelectionDAGISel::runOnMachineFunction(llvm::MachineFunction&) (/usr/lib/llvm-15/bin/../lib/libLLVM-15.so.1+0x167943c)
#14 0x0000ffff7ebc8fdc llvm::MachineFunctionPass::runOnFunction(llvm::Function&) (/usr/lib/llvm-15/bin/../lib/libLLVM-15.so.1+0x1258fdc)
#15 0x0000ffff7e9a05ec llvm::FPPassManager::runOnFunction(llvm::Function&) (/usr/lib/llvm-15/bin/../lib/libLLVM-15.so.1+0x10305ec)
#16 0x0000ffff7e9a71fc llvm::FPPassManager::runOnModule(llvm::Module&) (/usr/lib/llvm-15/bin/../lib/libLLVM-15.so.1+0x10371fc)
#17 0x0000ffff7e9a1014 llvm::legacy::PassManagerImpl::run(llvm::Module&) (/usr/lib/llvm-15/bin/../lib/libLLVM-15.so.1+0x1031014)
#18 0x0000ffff85d127bc clang::EmitBackendOutput(clang::DiagnosticsEngine&, clang::HeaderSearchOptions const&, clang::CodeGenOptions const&, clang::TargetOptions const&, clang::LangOptions const&, llvm::StringRef, llvm::Module*, clang::BackendAction, std::unique_ptr<llvm::raw_pwrite_stream, std::default_delete<llvm::raw_pwrite_stream>>) (/usr/lib/llvm-15/bin/../lib/libclang-cpp.so.15+0x18f27bc)
#19 0x0000ffff8600145c (/usr/lib/llvm-15/bin/../lib/libclang-cpp.so.15+0x1be145c)
#20 0x0000ffff84eb427c clang::ParseAST(clang::Sema&, bool, bool) (/usr/lib/llvm-15/bin/../lib/libclang-cpp.so.15+0xa9427c)
#21 0x0000ffff86994f00 clang::FrontendAction::Execute() (/usr/lib/llvm-15/bin/../lib/libclang-cpp.so.15+0x2574f00)
#22 0x0000ffff869249d8 clang::CompilerInstance::ExecuteAction(clang::FrontendAction&) (/usr/lib/llvm-15/bin/../lib/libclang-cpp.so.15+0x25049d8)
#23 0x0000ffff86a0c534 clang::ExecuteCompilerInvocation(clang::CompilerInstance*) (/usr/lib/llvm-15/bin/../lib/libclang-cpp.so.15+0x25ec534)
#24 0x0000aaaaca982e7c cc1_main(llvm::ArrayRef<char const*>, char const*, void*) (/usr/lib/llvm-15/bin/clang+0x12e7c)
#25 0x0000aaaaca981394 (/usr/lib/llvm-15/bin/clang+0x11394)
#26 0x0000ffff8660f0f8 (/usr/lib/llvm-15/bin/../lib/libclang-cpp.so.15+0x21ef0f8)
#27 0x0000ffff7e78c720 llvm::CrashRecoveryContext::RunSafely(llvm::function_ref<void ()>) (/usr/lib/llvm-15/bin/../lib/libLLVM-15.so.1+0xe1c720)
#28 0x0000ffff8660eae4 clang::driver::CC1Command::Execute(llvm::ArrayRef<llvm::Optional<llvm::StringRef>>, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char>>*, bool*) const (/usr/lib/llvm-15/bin/../lib/libclang-cpp.so.15+0x21eeae4)
#29 0x0000ffff865dddd0 clang::driver::Compilation::ExecuteCommand(clang::driver::Command const&, clang::driver::Command const*&, bool) const (/usr/lib/llvm-15/bin/../lib/libclang-cpp.so.15+0x21bddd0)
#30 0x0000ffff865de00c clang::driver::Compilation::ExecuteJobs(clang::driver::JobList const&, llvm::SmallVectorImpl<std::pair<int, clang::driver::Command const*>>&, bool) const (/usr/lib/llvm-15/bin/../lib/libclang-cpp.so.15+0x21be00c)
#31 0x0000ffff865f3d14 clang::driver::Driver::ExecuteCompilation(clang::driver::Compilation&, llvm::SmallVectorImpl<std::pair<int, clang::driver::Command const*>>&) (/usr/lib/llvm-15/bin/../lib/libclang-cpp.so.15+0x21d3d14)
#32 0x0000aaaaca980aac clang_main(int, char**) (/usr/lib/llvm-15/bin/clang+0x10aac)
#33 0x0000ffff7d4c7780 (/lib/aarch64-linux-gnu/libc.so.6+0x27780)
#34 0x0000ffff7d4c7858 __libc_start_main (/lib/aarch64-linux-gnu/libc.so.6+0x27858)
#35 0x0000aaaaca97e270 _start (/usr/lib/llvm-15/bin/clang+0xe270)
clang: error: clang frontend command failed with exit code 70 (use -v to see invocation)
Ubuntu clang version 15.0.6
Target: bpf
Thread model: posix
InstalledDir: /usr/bin
clang: note: diagnostic msg: 
********************

PLEASE ATTACH THE FOLLOWING FILES TO THE BUG REPORT:
Preprocessed source(s) and associated run script(s) are located at:
clang: note: diagnostic msg: /tmp/injection-4c830d.c
clang: note: diagnostic msg: /tmp/injection-4c830d.sh
clang: note: diagnostic msg: 

********************
make: *** [Makefile:29: disk-failure] Error 1
make: *** [docker-build-ebpf] Error 2
```
</p>
</detail>

Reply:
No worries if it does not work this was a small suggested improvement